core/services/directrequestocr/drlistener.go

```diff
@@ -0,0 +1,271 @@
+		case <-l.chStop:
+			l.shutdownWaitGroup.Done()
+			return
+		case <-l.mbOracleEvents.Notify():
+			for {
+				lb, exists := l.mbOracleEvents.Retrieve()
+				if !exists {
+					break
+				}
+				was, err := l.logBroadcaster.WasAlreadyConsumed(lb)
+				if err != nil {
+					l.logger.Errorw("Could not determine if log was already consumed", "error", err)
+					break
+				} else if was {
+					break
+				}
+
+				log := lb.DecodedLog()
+				if log == nil || reflect.ValueOf(log).IsNil() {
+					l.logger.Error("processOracleEvents: ignoring nil value")
+					break
+				}
+
+				switch log := log.(type) {
+				case *ocr2dr_oracle.OCR2DROracleOracleRequest:
+					l.shutdownWaitGroup.Add(1)
+					go l.handleOracleRequest(log, lb)
+				case *ocr2dr_oracle.OCR2DROracleOracleResponse:
+					l.shutdownWaitGroup.Add(1)
+					go l.handleOracleResponse(log, lb)
+				default:
+					l.logger.Warnf("Unexpected log type %T", log)
+				}
+			}
+		}
+	}
+}
+
+func (l *DRListener) handleOracleRequest(request *ocr2dr_oracle.OCR2DROracleOracleRequest, lb log.Broadcast) {
+	defer l.shutdownWaitGroup.Done()
+	l.logger.Infow("Oracle request received",
+		"requestId", fmt.Sprintf("%0x", request.RequestId),
+		"data", fmt.Sprintf("%0x", request.Data),
+	)
+
+	requestData := make(map[string]interface{})
+	requestData["requestId"] = formatRequestId(request.RequestId)
+	requestData["data"] = fmt.Sprintf("0x%x", request.Data)
+	meta := make(map[string]interface{})
+	meta["oracleRequest"] = requestData
+
+	runCloserChannel := make(chan struct{})
+	runCloserChannelIf, loaded := l.runs.LoadOrStore(formatRequestId(request.RequestId), runCloserChannel)
+	if loaded {
+		runCloserChannel, _ = runCloserChannelIf.(chan struct{})
+	}
+	ctx, cancel := utils.ContextFromChan(runCloserChannel)
+	defer cancel()
+
+	vars := pipeline.NewVarsFrom(map[string]interface{}{
+		"jobSpec": map[string]interface{}{
+			"databaseID":    l.job.ID,
+			"externalJobID": l.job.ExternalJobID,
+			"name":          l.job.Name.ValueOrZero(),
+		},
+		"jobRun": map[string]interface{}{
+			"meta":                  meta,
+			"logBlockHash":          request.Raw.BlockHash,
+			"logBlockNumber":        request.Raw.BlockNumber,
+			"logTxHash":             request.Raw.TxHash,
+			"logAddress":            request.Raw.Address,
+			"logTopics":             request.Raw.Topics,
+			"logData":               request.Raw.Data,
+			"blockReceiptsRoot":     lb.ReceiptsRoot(),
+			"blockTransactionsRoot": lb.TransactionsRoot(),
+			"blockStateRoot":        lb.StateRoot(),
+		},
+	})
+	run := pipeline.NewRun(*l.job.PipelineSpec, vars)
+	dbRequestID, err := l.pluginORM.CreateRequest(request.RequestId, time.Now(), &request.Raw.TxHash)
+	if err != nil {
+		l.logger.Errorf("Failed to create a DB entry for new request (ID: %v)", request.RequestId)
+		return
+	}
+	// TODO (bolek): Is Run() thread-safe?
+	_, err = l.pipelineRunner.Run(ctx, &run, l.logger, true, func(tx pg.Queryer) error {
+		l.markLogConsumed(lb, pg.WithQueryer(tx))
```

Comment:
I think we'd want to mark the log consumed as part of a db tx along with saving the result. Otherwise you could mark it complete but not actually save the result

Reply:
Log can be marked consumed pretty much as soon as we save the request to our DB (pluginORM.CreateRequest()). From that moment, every failure (in pipeline, saving the result, etc) will eventually be considered a timeout in processing. DON will produce a report without this particular node's observation.

I was also debating some kind of garbage collection/retry logic but I'm not convinced if that makes sense yet. Will revisit later.

Reply:
I see, for a given node its just best effort processing